adminpages/orders.php

```diff
@@ -385,10 +410,35 @@
 
 <?php if ( ! empty( $order ) ) { ?>
 
-	<?php if ( ! empty( $order->id ) ) { ?>
+	<?php if ( ! empty( $order->id ) ) { 
+		$refund_text = esc_html(
+			sprintf(
+				// translators: %s is the Order Code.
+				__( 'Refunding an order is permanent and can affect active users. Are you sure you want to refund order %s?', 'paid-memberships-pro' ),
+				str_replace( "'", '', $order->code )
+			)
+		);
+
+		$refund_nonce_url = wp_nonce_url(
+			add_query_arg(
+				[
+					'page'   => 'pmpro-orders',
+					'action' => 'refund_order',
+					'refund' => $order->id,
+					'order'  => $order->id
+				],
+				admin_url( 'admin.php' )
+			),
+			'refund_order',
+			'pmpro_orders_nonce'
+		);
+		?>
 		<h1 class="wp-heading-inline"><?php esc_html_e( 'Order', 'paid-memberships-pro' ); ?> #<?php echo esc_html( $order->id ); ?>: <?php echo esc_html( $order->code ); ?></h1>
 		<a title="<?php esc_attr_e( 'Print', 'paid-memberships-pro' ); ?>" href="<?php echo esc_url( add_query_arg( array( 'action' => 'pmpro_orders_print_view', 'order' => $order->id ), admin_url( 'admin-ajax.php' ) ) ); ?>" class="page-title-action" target="_blank" ><?php esc_html_e( 'Print', 'paid-memberships-pro' ); ?></a>
 		<a title="<?php esc_attr_e( 'Email', 'paid-memberships-pro' ); ?>" href="#TB_inline?width=600&height=200&inlineId=email_invoice" class="thickbox email_link page-title-action" data-order="<?php echo esc_html( $order->id ); ?>"><?php esc_html_e( 'Email', 'paid-memberships-pro' ); ?></a>
+		<?php if( pmpro_allowed_refunds( $order ) ) { ?>
+			<a title="<?php esc_attr_e( 'Refund', 'paid-memberships-pro' ); ?>" href="<?php echo 'javascript:pmpro_askfirst(\'' . esc_js( $refund_text ) . '\', \'' . esc_js( $refund_nonce_url ) . '\'); void(0);'; ?>" class="page-title-action"><?php esc_html_e( 'Refund', 'paid-memberships-pro' ); ?></a>
```

Comment:
Don't escape individual strings with `esc_js()`, use `wp_json_encode()` instead because it will handle all of the necessary escaping for you.

Then you would something like `attr="' . esc_js( $your_js ) . '"` and that would escape the attribute the way it needs to for the JS context (slightly differently from `esc_attr()`).

Doc on `esc_js()` has a note from Weston that really highlights why: https://developer.wordpress.org/reference/functions/esc_js/#comment-1747

Reply:
Will keep that in mind. This button was copied from existing code so figured I'd keep the same code going, but will change it up

Reply:
Done